Add Tag Functionality to the explorer

valentinpurrucker opened the pull request:
- add backend to map the paths of the items in the explorer to the assigned tags
- add new dialog to select tags, add new tags, update and delete tags

Closes #45 
Closes #48 
Closes #49 
Closes #50

Commit:
Add TagStore mock

45: Tagging

Task-Url: https://github.com/Jooshua/e4CompareFramework/issues/45

Commit:
Add FXML to CustomTreeCell

Is now fully customizable and supports tags.

45: Tagging

Task-Url: https://github.com/Jooshua/e4CompareFramework/issues/45

Commit:
Add tag backend using properties

- Rename TagStore to TagService
- Add ITagStore interface
- Add PropertiesTagStore

48: Tagging implement backend

Task-Url: https://github.com/Jooshua/e4CompareFramework/issues/48

Commit:
Improve tag service and store

48: Tagging implement backend

Task-Url: https://github.com/Jooshua/e4CompareFramework/issues/48

Commit:
Remove TODO

48: Tagging implement backend

Task-Url: https://github.com/Jooshua/e4CompareFramework/issues/48

Commit:
Add SerializableTagStore

48: Tagging implement backend

Task-Url: https://github.com/Jooshua/e4CompareFramework/issues/48

Commit:
Add a tag store mock and convenience function

48: Tagging implement backend

Task-Url: https://github.com/Jooshua/e4CompareFramework/issues/48

Commit:
Add tag search via toolbar

50: Tagging add filtering

Task-Url: https://github.com/Jooshua/e4CompareFramework/issues/50

Commit:
Make search case insensitive

50: Tagging add filtering

Task-Url: https://github.com/Jooshua/e4CompareFramework/issues/50

Commit:
Make get tag for tree element always NPE save

48: Tagging implement backend

Task-Url: https://github.com/Jooshua/e4CompareFramework/issues/48

valentinpurrucker committed:
Add user interface for adding tags

Task: #49

Commit:
Fix tagging backend loading problem

48: Tagging implement backend

Task-Url: https://github.com/Jooshua/e4CompareFramework/issues/48

valentinpurrucker committed:
Integrate Tag Pages into one

Moved init logic into Controller class
Renamed old files to match new dialog structure

Task: #49

Commit:
Add more functionality to the TagService

- Add function to move a tag
- Add function to update the color of a tag
- Remove old TagStores

48: Tagging implement backend

Task-Url: https://github.com/Jooshua/e4CompareFramework/issues/48

valentinpurrucker committed:
Merge remote-tracking branch 'origin/feature/task_45-Tagging_baseline'
into feature/task_49-Tag_Editing

Task: #49

valentinpurrucker committed:
Add tag update functionality in UI

- remove unnecessary fxml stuff (id, text)
- add some javadoc

Task: #49

Commit:
Improve tagging look and feel

- A tag can create its own ui representation now to remove duplicate
code
- Update javafx version numbers to avoid console warnings
- Add button for selecting and deselecting all tags
- TagService now only hands out sorted lists to have consistent ui

45: Tagging

Task-Url: https://github.com/Jooshua/e4CompareFramework/issues/45

Commit:
Improve tagging look and feel

- A tag can create its own ui representation now to remove duplicate
code
- Update javafx version numbers to avoid console warnings
- Add button for selecting and deselecting all tags
- TagService now only hands out sorted lists to have consistent ui

45: Tagging

Task-Url: https://github.com/Jooshua/e4CompareFramework/issues/45

timbittner commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/controller/ProjectExplorerViewController.java:
```diff
@@ -348,8 +368,25 @@ public void rename(@UIEventTopic(E4CEventTable.EVENT_RENAME_PROJECT_EXPLORER_ITE
-			filter = (String) o;
+			String filterString = (String) o;
+
+			if (filterString.contains(TagService.TAG_PREFIX)) {
```
Documentation

timbittner commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/controller/ProjectExplorerViewController.java:
```diff
@@ -348,8 +368,25 @@ public void rename(@UIEventTopic(E4CEventTable.EVENT_RENAME_PROJECT_EXPLORER_ITE
 	@Inject
 	@Optional
 	public void filter(@UIEventTopic(E4CEventTable.EVENT_FILTER_CHANGED) Object o) {
+		filterTags.clear();
```
are you sure that you want to always clear the filterTags? Even if you dont know anything about "o" ?

Jooshua replied:
yes

timbittner commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/tagging/Tag.java:
```diff
@@ -0,0 +1,108 @@
+import javafx.scene.paint.Color;
+import javafx.scene.shape.Circle;
+
+public class Tag implements Serializable {
```
Javadoc and reconsider if implementing the Serializable interface is still necessary with the current approach

timbittner commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/tagging/store/ITagStore.java:
```diff
@@ -0,0 +1,34 @@
+	
+	/**
+	 * Store the available tags
+	 * @param availableTags
```
Javadoc does not further explain parameter availableTags

timbittner commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/tagging/store/ITagStore.java:
```diff
@@ -0,0 +1,34 @@
+	 * Load a map of tags for each file
+	 * @return map with tags
+	 */
+	public Map<String, List<Tag>> loadTagMap(List<Tag> availableTags);
```
Javadoc missing parameter, also why is it required

timbittner commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/tagging/store/ITagStore.java:
```diff
@@ -0,0 +1,34 @@
+	 * Store a map of tags for each file
+	 * @param tagMap
+	 */
+	public void storeTagMap(Map<String, List<Tag>> tagMap);
```
Javadoc missing parameter, also why is it required

timbittner commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/tagging/TagService.java:
```diff
@@ -0,0 +1,292 @@
+		}
+	}
+
+	private List<Tag> saveGetListFromMap(String path) {
```
Javadoc. What exactly is saving? put? Did you mean safe?

timbittner commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/tagging/Tag.java:
```diff
@@ -0,0 +1,108 @@
+
+		Tag tag = (Tag) obj;
+
+		return (tag.name == this.name && tag.colorString == this.colorString);
```
Tags with the same name should be considered equal as the name is used as a unique identifier (might also solve one of my mentioned bugs)

timbittner commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/tagging/dialog/TagDialog.java:
```diff
@@ -0,0 +1,99 @@
+	public TagDialog(IEclipseContext context, TagService tagService, ServiceContainer services,
+			List<FileTreeElement> selectedElements) {
+		alert = new Alert(AlertType.NONE);
+		alert.setTitle("Manage Tags");
```
Icon probably here?

timbittner commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/tagging/dialog/TagDialog.java:
```diff
@@ -0,0 +1,99 @@
+
+	/** show the dialog and wait for the users input */
+	public void open() {
+		alert.showAndWait().filter(response -> response == ButtonType.OK).ifPresent(response -> {
```
Make sure that pressing cancel does cancel actions taken in the wizard (creating, deleting  tags etc)

valentinpurrucker replied:
~~In my opinion there are two approaches~~

1. determine what has changed and apply those changes to the tagService by calling addTag when a new tag has been added and delete when a tag has been deleted
2. ~~my favorite approach :D just work directly one the available tag list, since all the methods of the service do the same and we even get the correct reference when calling stuff like getavaiable tags. atm i don't see the downside of that.
I started implementing the basic idea of that in https://github.com/Jooshua/e4CompareFramework-team2/pull/53/commits/b12c29fdf2f5ce0b242447fc353d64bb9060d373 in the TagDialog~~
Edit: dumb idea, obviously we need to clear all the entries in the file when deleting tags 🙈

timbittner reviewed (changes requested):
Great work guys! I really enjoy your implementation :)
I found a few issues while playing around that you should fix. 

 

- [x] Manage Tags missing icon
- [x] Wizard Popup missing E4C icon (see new file dialog)
- [x] TagService  should be in ServiceContainer and moved to core package with dependencies
- [x] (ITagStore not in interface package)
- [x] Can Create the same tag with a different color and use it in the current session. After restarting only one tag will be applied _(Comment: lets just ensure that the user cant create those via the ui)_
- [x] Invalid Tags may be created that are stored but not reapplied: tagName=":Invalid1" ==> Forbid ':' _(Comment: lets just ensure that the user cant create those via the ui)_
- [x] Tags do not display in ListView but may be edited (enable listview) _(Comment: cant reproduce)_
- [x] Deleting a Tag and then aborting the dialog deletes the tag anyway but does not update the View
- [x] Add Action to ToolBar (consider Handler, canExecute!)

Commit:
First changes due to review

- Add missing documentation
- Fix Tag equal function
- Move Tag classes into core and add service to service container


45: Tagging

Task-Url: https://github.com/Jooshua/e4CompareFramework/issues/45

Jooshua commented:
> 
> 
> Great work guys! I really enjoy your implementation :)
> I found a few issues while playing around that you should fix.
> 
>     * [ ]  Manage Tags missing icon
> 
>     * [ ]  Wizard Popup missing E4C icon (see new file dialog)
> 
>     * [x]  TagService  should be in ServiceContainer and moved to core package with dependencies
> 
>     * [x]  (ITagStore not in interface package)
> 
>     * [ ]  Can Create the same tag with a different color and use it in the current session. After restarting only one tag will be applied
> 
>     * [ ]  Invalid Tags may be created that are stored but not reapplied: tagName=":Invalid1" ==> Forbid ':'
> 
>     * [ ]  Tags do not display in ListView but may be edited (enable listview)
> 
>     * [ ]  Deleting a Tag and then aborting the dialog deletes the tag anyway but does not update the View
> 
>     * [ ]  Add Action to ToolBar (consider Handler, canExecute!)

I cannot reproduce the list/flat view problem.
@valipu As I would consider the remaining issues UI could you please look into them.

valentinpurrucker committed:
Changes due to PR

- manage session specific tag list
- add cancel btn
- add error message for wrong tag names

Task: #45

valentinpurrucker committed:
Merge branch 'feature/task_45-Tagging_baseline' of git@github.com:Jooshua/e4CompareFramework.git into feature/task_45-Tagging_baseline

valentinpurrucker committed:
Changes due to PR

valentinpurrucker committed:
Add Tag Manager to Toolbar

Task: #45

valentinpurrucker commented:
Only thing that bothers me right now -> the dialog doesn't open if the button in the toolbar is pressed, although the execute method is executed :(

Update: resolved. this was due to an issue with the tagservice not being injected when calling the handler manually

valentinpurrucker committed:
Changes due to PR

- Fix not opening dialog bug
- fix assignment of new session tags
- some code cleanup

Task: #45

valentinpurrucker committed:
Remove TagService injection into handler

- tagservices is provided in services

Task: #45

timbittner commented on de.tu_bs.cs.isf.e4cf.core/src/de/tu_bs/cs/isf/e4cf/core/util/tagging/ITagStore.java:
```diff
@@ -1,33 +1,33 @@
 	public Map<String, List<Tag>> loadTagMap(List<Tag> availableTags);
 	
 	/**
 	 * Store a map of tags for each file
```
Just from the interface javadoc it is still unclear what exactly these functions should do. E.g. what is the string part of this map.

As an interface documentation should definitely be more detailed.

timbittner commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/ui/view/TagView.fxml:
```diff
@@ -10,34 +10,26 @@
-	spacing="4" xmlns="http://javafx.com/javafx/8.0.171"
-	xmlns:fx="http://javafx.com/fxml/1"
-	fx:controller="de.tu_bs.cs.isf.e4cf.parts.project_explorer.controller.TagViewController">
+<VBox maxHeight="-Infinity" maxWidth="-Infinity" minHeight="-Infinity" minWidth="-Infinity" prefWidth="600.0" spacing="4" xmlns="http://javafx.com/javafx/11.0.1" xmlns:fx="http://javafx.com/fxml/1" fx:controller="de.tu_bs.cs.isf.e4cf.parts.project_explorer.controller.TagViewController">
```
JavaFX 11 leading to red errors in console

timbittner reviewed (changes requested):
Great work. All bugs seem fixed.

The interface still lacks documentation for my taste, as interfaces should be really extensive in their JavaDoc.

valentinpurrucker committed:
Fix FXML version

Task: #45

Commit:
Improve ITagStore documentation

45: Tagging

Task-Url: https://github.com/Jooshua/e4CompareFramework/issues/45

kyshidar commented on de.tu_bs.cs.isf.e4cf.core/src/de/tu_bs/cs/isf/e4cf/core/util/tagging/TagService.java:
```diff
@@ -0,0 +1,290 @@
+	 * 
+	 * @param tag
+	 */
+	public void delteAvailableTag(Tag tag) {
```
delete spelled wrong

Commit:
Fix spelling

45: Tagging

Task-Url: https://github.com/Jooshua/e4CompareFramework/issues/45
